state/keys.go

```diff
@@ -3,85 +3,55 @@
-	Allocate PermissionBit = 1
-	Write    PermissionBit = 2
-)
+	Allocate
+	Write
 
-// We use a bitset (8 bits) in the Action and Auth marshal/unmarshaling.
-type PermissionsBits set.Bits
+	None Permissions = 0
+	All              = Read | Allocate | Write
+)
 
-// Bit representation of Read/Allocate/Write
-type PermissionBit int
+var errUnexpectedLength = errors.New("permissions should be exactly one byte")
 
 // StateKey holds the name of the key and its permission (Read/Allocate/Write). By default,
 // initialization of Keys with duplicate key will not work. And to prevent duplicate
 // insertions from overriding the original permissions, use the Add function below.
-type Keys map[string]*Permissions
+type Keys map[string]Permissions
 
 // All acceptable permission options
-type Permissions struct {
-	Read     bool
-	Allocate bool
-	Write    bool
-}
+type Permissions byte
 
 // Transactions are expected to use this to prevent
 // overriding of key permissions
-func (k Keys) Add(name string, permission *Permissions) {
-	_, exists := k[name]
-	if !exists {
-		k[name] = permission
-	} else {
-		// Transaction's permissions are the union of all
-		// state keys from both Actions and Auth
-		if permission.Read {
-			k[name].Read = true
-		}
-		if permission.Allocate {
-			k[name].Allocate = true
-		}
-		if permission.Write {
-			k[name].Write = true
-		}
-	}
+func (k Keys) Add(name string, permission Permissions) {
+	// Transaction's permissions are the union of all
+	// state keys from both Actions and Auth
+	k[name] |= permission
+}
+
+// Has returns true if [p] has all the permissions that are contained in require
+func (p Permissions) Has(require Permissions) bool {
+	return require&^p == 0
 }
 
 // Populates bitset using the values from Permissions struct
 // We don't need to worry about adding invalid permission bits
 // since our struct is defined with all valid ones
 func (p Permissions) ToBytes() []byte {
-	permBitSet := set.NewBits()
-	if p.Read {
-		permBitSet.Add(int(Read))
-	}
-	if p.Allocate {
-		permBitSet.Add(int(Allocate))
-	}
-	if p.Write {
-		permBitSet.Add(int(Write))
-	}
-	return permBitSet.Bytes()
+	return []byte{byte(p)}
 }
 
 // Populates Permissions struct with the permissions bytes
 // Any permission bits that were not defined will be dropped
-// By default, all fields in Permissions are false
-func FromBytes(permissionBytes []byte) *Permissions {
-	var permissions Permissions
-	permissionBits := set.BitsFromBytes(permissionBytes)
-	if permissionBits.Contains(int(Read)) {
-		permissions.Read = true
+func FromBytes(permissionBytes []byte) (*Permissions, error) {
+	if len(permissionBytes) != 1 {
+		return nil, errUnexpectedLength
 	}
-	if permissionBits.Contains(int(Allocate)) {
-		permissions.Allocate = true
-	}
-	if permissionBits.Contains(int(Write)) {
-		permissions.Write = true
-	}
-	return &permissions
+
+	permissionBits := Permissions(permissionBytes[0]) & All
+	return &permissionBits, nil
 }
```

Comment:
I'm not really sure these are needed anymore... They feel a bit heavy now that `Permissions` is already a `byte`